src/class/audio/audio_device.c

```diff
@@ -2038,81 +2041,94 @@ bool audiod_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint3
+    audio->fb_float_val = (float)f_s / f_m * (1UL << (16 - audio->fb_n_frames_shift));
+  }else
+  {
+    audio->fb_param_factor_N = f_s;
+    audio->fb_param_factor_D = f_m;
   }
 
-  audio->fb_power_of_two_val = 16 - tu_log2(audio->fb_n_frames * f_m / f_s);
-#endif
-
-#if CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE_FLOAT
-  audio->fb_float_val = (float)f_s / audio->fb_n_frames / f_m * (1 << 16);
-#endif
+  // Minimal/Maximum value in 16.16 format for full speed (1ms per frame) or high speed (125 us per frame)
+  uint32_t const frame_div = (TUSB_SPEED_FULL == tud_speed_get()) ? 1000 : 8000;
+  audio->fb_val_min = (f_s/frame_div - 1) << 16;
+  audio->fb_val_max = (f_s/frame_div + 1) << 16;
 
-#if CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE_FIXED_POINT
+  return true;
+}
 
-  // f_s max is 2^19-1 = 524287 Hz
-  // f_m max is 2^29/(1 ms * n_frames) for full speed and 2^29/(125 us * n_frames) for high speed, this means for f_m fitting in an uint32, n_frames must be < 125 for full speed and < 1000 for high speed (1000 does not fit into uint8 so the maximum possible value cannot even be reached by UAC2 - we don't need to check for it)
+uint32_t tud_audio_feedback_update(uint8_t func_id, uint32_t cycles)
+{
+  audiod_function_t* audio = &_audiod_fct[func_id];
+  uint32_t feedback;
 
-  if ((f_s > (2^19)-1) || (TUSB_SPEED_FULL == tud_speed_get() && (audio->fb_n_frames > 125)))
+  switch (audio->fb_compute_method)
   {
-    // If this check fails, not every thing is lost - you need to re-evaluate the scaling factors of the parameters such that the numbers fit into uint64 again. feedback = n_cycles * S_c / (n_frames * S_n) * f_s * S_s / (f_m * S_m). In the end S_c*S_s / (S_n * S_m) = 2^16 for a 16.16 fixed point precision. If you find something, define your own function of tud_audio_set_feedback_params_fm_fs() and audiod_sof() and use your values
-    TU_LOG2("  FEEDBACK_DETERMINATION_MODE_FIXED_POINT not possible!\r\n"); TU_BREAKPOINT(); return false;
-  }
+    case FEEDBACK_COMPUTE_POWER_OF_2:
+      feedback = cycles << audio->fb_power_of_two_val;
+    break;
 
-  audio->fb_param_factor_N = (uint64_t)f_s << 13;
-  audio->fb_param_factor_D = (uint64_t)f_m * audio->fb_n_frames;
-#endif
+    case FEEDBACK_COMPUTE_FLOAT:
+      feedback = (uint32_t) ((float) cylces * audio->fb_float_val);
+    break;
 
-  audio->fb_val_min = ((TUSB_SPEED_FULL == tud_speed_get() ? (f_s/1000) : (f_s/8000)) - 1) << 16;   // Minimal value in 16.16 format for full speed (1ms per frame) or high speed (125 us per frame)
-  audio->fb_val_max = ((TUSB_SPEED_FULL == tud_speed_get() ? (f_s/1000) : (f_s/8000)) + 1) << 16;   // Maximum value in 16.16 format
+    case FEEDBACK_COMPUTE_FIXED:
+    {
+      uint64_t fb64 = (((uint64_t) cycles) * audio->fb_param_factor_N) << (16 - audio->fb_n_frames_shift);
```

Comment:
since n_frame is power of 2, I just store the sample & mclk freq and do the shift `(16 - (bInterval-1))`. The shift is fast enough, and we don't have to store 2x 64-bit. 

![image](https://user-images.githubusercontent.com/249515/168334764-39199458-f1a5-46da-ba16-a700a50887fa.png)